File: reference/dc.py
def dc(highs, lows, period):
	upper = []
	lower = []
	middle = []
	for i in range(len(highs)):
		if i < period - 1:
			upper.append(None)
			lower.append(None)
			middle.append(None)
		else:
			upper.append(max(highs[i-period+1:i+1]))
			lower.append(min(lows[i-period+1:i+1]))
			middle.append((upper[-1] + lower[-1]) / 2)
	return upper, lower, middle

def dc2(highs, lows, period):
	
	upper = []
	lower = []
	middle = []

	for i in range(period-1, len(highs)):
		upper.append(max(highs[i-period+1:i+1]))
		lower.append(min(lows[i-period+1:i+1]))
		middle.append((upper[-1] + lower[-1]) / 2)

	return upper, lower, middle

File: src/core/extract/dc.ts


/**
 * @param highs 
 * @param lows 
 * @param period 
 * @returns [upper, middle, lower]
 */
export async function dc(highs: number[], lows: number[], period: number) {

    const upper: number[] = []
    const lower: number[] = []
    const middle: number[] = []

    for (let i = period - 1; i < highs.length; i++) {
        upper.push(Math.max(...highs.slice(i - period + 1, i + 1)))
        lower.push(Math.min(...lows.slice(i - period + 1, i + 1)))
        middle.push((upper[upper.length - 1] + lower[lower.length - 1]) / 2)
    }

    return [upper, middle, lower]
}